libobs/obs-output.c

```diff
@@ -1989,6 +1993,85 @@ static void discard_unused_audio_packets(struct obs_output *output,
 		discard_to_idx(output, idx);
 }
 
+static bool purge_encoder_group_keyframe_data(obs_output_t *output, size_t idx)
+{
+	struct keyframe_group_data *data =
+		&output->keyframe_group_tracking.array[idx];
+	uint32_t modified_count = 0;
+	for (size_t i = 0; i < MAX_OUTPUT_VIDEO_ENCODERS; i++) {
+		if (data->seen_on_track[i] != KEYFRAME_TRACK_STATUS_NOT_SEEN)
+			modified_count += 1;
+	}
+
+	if (modified_count == data->required_tracks) {
+		da_erase(output->keyframe_group_tracking, idx);
+		return true;
+	}
+	return false;
+}
+
+/* Check whether keyframes are emitted from all grouped encoders, and log
+ * if keyframes haven't been emitted from all grouped encoders. */
+static void
+check_encoder_group_keyframe_alignment(obs_output_t *output,
+				       struct encoder_packet *packet)
+{
+	size_t idx = 0;
+	struct keyframe_group_data insert_data = {0};
+
+	if (!packet->keyframe || packet->type != OBS_ENCODER_VIDEO ||
+	    !packet->encoder->encoder_group)
+		return;
+
+	for (; idx < output->keyframe_group_tracking.num;) {
+		struct keyframe_group_data *data =
+			&output->keyframe_group_tracking.array[idx];
+		if (data->pts > packet->pts)
+			break;
+		if (data->group_id !=
+		    (uintptr_t)packet->encoder->encoder_group) {
+			idx += 1;
+			continue;
+		}
+
+		if (data->pts < packet->pts) {
+			if (data->seen_on_track[packet->track_idx] ==
+			    KEYFRAME_TRACK_STATUS_NOT_SEEN) {
+				blog(LOG_WARNING,
```

Comment:
Just a note for the future: LOG_WARNING may be too noisy. We'll have to evaluate as this gets used in the wild if this should be LOG_DEBUG.

Reply:
This message doesn't get output normally, it's basically considered an error if this error message is output, so it's probably important and makes sense to output it as a general log warning in this case. Otherwise we'll never know about it from user logs.

Reply:
(Or rather considered a warning that something is going awry and should be examined. Stream should still technically work I believe)

Reply:
Huh, Ryan sent me a log file where the user had dynamic bitrate enabled (which I assume isn't supported) and it *does* get very verbose with multiple messages per second. Hmm.

Reply:
Well I think that's being patched so I think it'll be fine for now. If we have any issues with logging we can adjust it later.

Reply:
My comment was mostly, "Let's be mindful of this and keep an eye out as this code makes it into use in the wild." I wanted it noted that I raised the concern as something to keep an eye on, but not a blocker at this time.